### ISSUE TITLE
How to train model based on existing weights?

### ISSUE THREAD

**Issue by tino926** [ultralytics/yolov5#4268]
## ❔Question
I have already trained a model. Then I try to improve it with different parameters based on the trained weights. 
However, when I use --weights parameter, yolov5 just finished immediately because all the epochs are done.

These are the commands I used
`python train.py --project proj --img 480 --batch-size 32 --epoch 300 --data ./data/abc.yaml --cfg ./models/abc.yaml --hyp data/hyp.scratch.yaml `

`python train.py --project proj --img 480 --batch-size 32 --epoch 600 --data ./data/abc.yaml --cfg ./models/abc.yaml --hyp data/hyp.finetune.yaml --weights proj/exp/weights/last.pt`

What is the proper command I should use?

**Comment by glenn-jocher** [comment]
@tino926 👋 Hello! Thanks for asking about **resuming training**. YOLOv5 🚀 Learning Rate (LR) [schedulers](https://github.com/ultralytics/yolov5/blob/9f3a388cea66ca71b9bbe057929eed8fbc920a7e/train.py#L109-L142) follow predefined LR curves for the fixed number of `--epochs` defined at training start (`default=300`), and are designed to fall to a minimum LR on the final epoch for best training results. For this reason you can not modify the number of epochs once training has started.

<img width="850" alt="LR Curves" src="https://user-images.githubusercontent.com/26833433/119332195-8e113a80-bc88-11eb-8eb7-7494ff1d5c01.png">

If your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.run --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck and let us know if you have any other questions!

**Comment by tino926** [comment]
Thanks for your quick reply.
I have another question. Can I "start training" from a partially trained model?

That says, 
If in the first exp, I set --epoch 600. During the training I found it hard to get better result at around 100th epoch and I manually stop it. 

In the second exp, I want to start a fresh training with different parameters but use the run/exp/weights/last.pt as the initial weights. How can I do that?

**Comment by glenn-jocher** [comment]
@tino926 YOLOv5 🚀 training checkpoints should be 4X the size of final trained checkpoints, as they carry not just a FP16 model, but a FP16 [EMA](https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage) and an FP32 optimizer of the same size as the model (each model parameter has its own FP32 gradient saved within the optimizer). Final checkpoints contain only an FP16 model, with the EMA and optimizer both stripped after the final epoch of training:
https://github.com/ultralytics/yolov5/blob/75feeb797c4a9553f4274860c6c278f1fc628f60/train.py#L435-L439

The `strip_optimizer()` function updates the checkpoint dictionary value for `model`, replacing it with `ema`, and sets `ema` and `optimizer` keys to `None`, which will reduce the checkpoint size by 3/4.
https://github.com/ultralytics/yolov5/blob/75feeb797c4a9553f4274860c6c278f1fc628f60/utils/general.py#L495-L509

Final trained checkpoints after `strip_optimizer()` should match the README table sizes below.
<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

**Comment by alicera** [comment]
yolov5s.pt is based on COCO data.
I test 1000 images and the detect result is not good with yolov5s.pt.
I add the 1000 images to train.

method1
Using coco + 1000 images to train the model.

method2 
Using 1000 images to train the yolov5s.pt model.
python train.py --weights path/to/yolov5s.pt   

expect 
Get the good result on 1000 images and keep the coco mAP accuracy.

Which method is better?

**Comment by glenn-jocher** [comment]
@alicera results with vary by use case, you should experiment with your custom dataset and deployment scenario to determine the best workflow for your problem.

**Comment by tino926** [comment]
@glenn-jocher ,
Similar question here. 
You have suggested that, one can set epochs=300 at first, and then try more epoches:
> Epochs. Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.

For "train longer", do you suggest a fresh training? Or train with the best weights we got from the first shorter training?
If we train with the weights from the first training, should we modify the training parameters?

This is what I've got when I try to train with existing weights:
First I train with 300 epochs:
```
python train.py --project r6 --epoch 300 --data ./data/d6.yaml --cfg ./models/yolov5ss2.yaml \
                         --hyp data/hyp.scratch.yaml 
```
and got
![image](https://user-images.githubusercontent.com/2639399/129124152-196f6c4d-6e7d-4865-b564-8e262f81282d.png)

It seems that I can get better scores with more epochs. Then I try to train with more epochs, but avoid the first 300 epochs, so I did this: make a new hyper.scratch_2.yaml from hyp.scratch.yaml by changing the value of `lr0` form 0.01 to 0.002 and use the command
```
python train.py --project r6 --epoch 300 --data ./data/d6.yaml --cfg ./models/yolov5ss2.yaml \
                         --hyp data/hyp.scratch_0.002.yaml \
                         --weights r6/exp/weights/best.pt 
```

This is the training result (gray line):
![image](https://user-images.githubusercontent.com/2639399/129125266-7404610a-85ea-4b6e-bd75-36fa1dbd5eae.png)

All metrics seem better in the second training, except for the precision. Do you have some suggestion? A better way to modify the hyper_parameter.yaml or just spend times to train from the beginning with larger epochs?

**Comment by glenn-jocher** [comment]
@tino926 that all looks good, yes your method of starting from best.pt is fine. You should be able to reproduce the same results with your second hyp file training 600 epochs in one training as well (from scratch).

P and R are really irrelevant as they evaluate at a single confidence, you should focus on mAP which is a function of all confidences.

**Comment by glenn-jocher** [comment]
@tino926 also overfitting is determined by your val losses, which are not displayed in your issue, training losses are irrelevant for determining overfitting.

**Comment by tino926** [comment]
@glenn-jocher, 
Thanks for your replay. 
I've tried a another training with `--weights r6/exp/weights/best.pt ` and got similar result.
Now I am just curious about why in the second training, the precision value starts at `~0.415` whereas it ends at `~0.435` in the first training. I guess it is because of momentum or some warmup parameters. 
I am now using yolov5 4.0, I'll later upgrade to 5.0 and try to figure out this.
